llama.cpp

```diff
@@ -7783,6 +7783,73 @@ void llama_sample_typical(struct llama_context * ctx, llama_token_data_array * c
     }
 }
 
+void llama_sample_entropy(struct llama_context * ctx, llama_token_data_array * candidates_p, float min_temp, float max_temp, float exponent_val) {
+    const int64_t t_start_sample_us = ggml_time_us();
+
+    llama_sample_softmax(ctx, candidates_p);
+
+    // Calculate entropy of the softmax probabilities
+    float entropy = 0.0f;
+    for (size_t i = 0; i < candidates_p->size; ++i) {
+        float prob = candidates_p->data[i].p;
+        if (prob > 0.0f) { // Ensure no log(0)
+            entropy -= prob * logf(prob);
+        }
+    }
+
+    // Calculate maximum possible entropy
+    float max_entropy = -logf(1.0f / candidates_p->size);
+
+    // Guard against division by zero
+    if (max_entropy == 0.0f) {
```

Comment:
The only way `max_entropy` can be zero is if we have a single candidate. If that's the case, then you can return right there because the code below does nothing to the probability of the single candidate. Or am I missing something?

Reply:
You are right. I've moved the check earlier in the code and return immediately in case of single candidate. In fact, it seems if there's only 1 candidate, the `softmax` function can be skipped as well.